packages/flutter/lib/src/widgets/spell_check.dart

```diff
@@ -0,0 +1,345 @@
+  });
+
+  /// The service used to fetch spell check results for text input.
+  final SpellCheckService? spellCheckService;
+
+  /// The handler used to mark misspelled words in text input and display
+  /// a menu of the replacement suggestions for these misspelled words.
+  final SpellCheckSuggestionsHandler? spellCheckSuggestionsHandler;
+
+  /// The most up-to-date spell check results for text input.
+  ///
+  /// These results will be updated by the
+  /// [spellCheckService] and used by the [spellCheckSuggestionsHandler] to
+  /// build the [TextSpan] tree for text input and menus for replacement
+  /// suggestions of misspelled words.
+  SpellCheckResults? spellCheckResults;
+
+  /// Configuration that indicates that spell check should not be run on text
+  /// input and/or spell check is not implemented on the respective platform.
+  static SpellCheckConfiguration disabled = SpellCheckConfiguration();
+}
+
+/// Determines how misspelled words are indicated in text input and how
+/// replacement suggestions for misspelled words are displayed via menu.
+mixin SpellCheckSuggestionsHandler {
+  /// Builds the [TextSpan] tree given the current state of the text input and
+  /// spell check results.
+  TextSpan buildTextSpanWithSpellCheckSuggestions(
+    TextEditingValue value,
+    bool composingWithinCurrentTextRange,
+    TextStyle? style,
+    SpellCheckResults spellCheckResults
+  );
+}
+
+/// The handler used by default for spell checking text input.
+///
+/// Any widget may use this handler to build a [TextSpan] tree with
+/// [SpellCheckResults] indicated in a style based on the platform by calling
+/// `buildTextSpanWithSpellCheckSuggestions(...)` with an instance of this
+/// class.
+///
+/// See also:
+///
+///  * [SpellCheckSuggestionsHandler], the handler that this implements and
+///    may be overriden for use by [EditableText].
+///  * [EditableText], which uses this handler to display spell check results
+///     by default if spell check is enabled.
+class DefaultSpellCheckSuggestionsHandler with SpellCheckSuggestionsHandler {
+  /// Creates a handler to use for spell checking text input based on the
+  /// provided platform.
+  DefaultSpellCheckSuggestionsHandler(TargetPlatform platform) {
+    switch (platform) {
+      case TargetPlatform.iOS:
+      case TargetPlatform.macOS:
+        misspelledTextStyle = cupertinoMisspelledTextStyle;
+        break;
+      case TargetPlatform.android:
+      case TargetPlatform.fuchsia:
+      case TargetPlatform.linux:
+      case TargetPlatform.windows:
+        misspelledTextStyle = materialMisspelledTextStyle;
+        break;
+    }
+  }
+
+  /// The text style by which misspelled words will be indicated in the
+  /// [TextSpan] tree based on platform.
+  late final TextStyle misspelledTextStyle;
+
+  /// The style used to indicate misspelled words on Android.
+  static final TextStyle materialMisspelledTextStyle = const TextStyle(
```

Comment:
The fact that this file in the widgets library needs to know these details about Material and Cupertino is making me think that there might be a better way to organize this.  I feel like if you put the Material stuff in the Material library and the Cupertino stuff in the Cupertino library and then pass that information into the widgets library via EditableText, it would end up working out.

  * The Cupertino library gets cupertinoMisspelledTextStyle.
  * The Material library gets materialMisspelledTextStyle.
  * Material *also* gets something that chooses the style based on that platform.  That logic goes in Material because Material knows about both itself Cupertino.
  * The widgets library retains the core logic from this class.
  * TextField passes in Material's platform-dependent style to EditableText.
  * CupertinoTextField passes in its Cupertino-specific style to EditableText.

That means that if you use a TextField on any platform, it will (by default) give you the correct platform-specific misspelling style.  However if you use CupertinoTextField, it will always give you the Cupertino misspelling style, even if you're using Android.  I think that's ok, and it's consistent with Flutter's architectural choices.  Currently, if you put a CupertinoTextField on an Android device, you'll get the iOS text selection toolbar.

Reply:
That is a lot of parameters.  I'm on board with putting all of that in one class :+1: